### ISSUE TITLE
connect() segmentation fault

### ISSUE THREAD

**Issue by nyankers** [dworkin/dgd#77]
A segmentation fault happens when connecting in an atomic function via `connect()` per how Cloud Server does things if you happen to forget to write one of the user commands. I was able to reproduce this on updated code with the following user object (I put it under `~System/~` just to dodge any possible access rights issues).

```
# include <kernel/user.h>

inherit "~/lib/user";

void create()
{
    connect("google.com", 80);
}

int receive_message(string str) {
    return MODE_NOCHANGE;
}
```

(Note the lack of a `login()` function, woops!)

Just call `create()` on the above and you'll get a segmentation fault.

It seems to be failing in comm.cpp line 828, my guess without doing further analysis is that maybe the object got cleaned by the rollback, but something in communications thinks there's still work to do.

**Comment by dworkin** [comment]
Should be fixed by ff823f9.

**Comment by nyankers** [comment]
Looks like it's fixed.

Thanks, as usual! <3

**Comment by nyankers** [comment]
Also, since I'm playing around with `connect()` and `send_message()` for outgoing connections, I found out that the Cloud Server `login()` function hook actually cannot send any messages to the outgoing connection directly, unlike when the hook is called for an incoming connection. Instead, they have to be put behind a `call_out()`, or the below error will be raised (and incidentally, will cause the `connect()` to fail on Cloud Server, since it's an uncaught exception).

```
Oct 24 17:46:06 ** Output channel closed [atomic]
                 /usr/System/sys/test
   35   connect     /kernel/lib/user
                 /kernel/obj/binary#134
   32   connect     /kernel/obj/binary
  197   set_user    /kernel/lib/connection
                 /usr/System/sys/test
   21   login       /usr/System/sys/test
                 /kernel/obj/binary#134
  248   message     /kernel/lib/connection
Oct 24 17:46:06 ** Output channel closed [caught]
                        /kernel/obj/telnet#128
   53   receive_message    /kernel/obj/telnet
  233   receive_message    /kernel/lib/connection
                        /kernel/obj/user#129
  247   receive_message    /kernel/obj/user
                        /kernel/obj/wiztool#130
   35   input              /kernel/obj/wiztool
   99   process            /kernel/obj/wiztool
  942 * cmd_code           /kernel/lib/wiztool
                        /usr/admin/_code
   10   exec               /usr/admin/_code
                        /usr/System/sys/test
   15   create             /usr/System/sys/test
```

(Above happened when trying to do `query_conn()->message("GET /")`, the object code is similar to but more complete than the above, I can provide it if desired.)

It's not difficult to code around this, but in case this is unintended behavior, I figured I'd bring it to your attention.

My assumption is that this is happening because whereas incoming connections have been set up before calling the appropriate driver functions, outgoing connections can't be set up until the frame ends.

**Comment by dworkin** [comment]
> My assumption is that this is happening because whereas incoming connections have been set up before calling the appropriate driver functions, outgoing connections can't be set up until the frame ends.

Correct.  In this case, `login` is called before the connection is actually established, and no messages can be sent before that.  You can send messages from `connected`. Your confusion is understandable, and points towards a forthcoming improvement in the cloud library.

**Comment by dworkin** [comment]
The latest cloudlib update will delay the call to `login` until the connection is established, and `connected` is no longer called in the user object.  `login` is still called with `nil` as an argument.

**Comment by dworkin** [comment]
It also updates `HTTP` to what should be some form of temporary stability. I've decided not to bother with a full HTTP server for now (which is a pretty big project) and to just provide the plumbing. That will do for the minimal web server that is already included, as well as future REST services.

**Comment by nyankers** [comment]
After some git wizardry and a whole bunch of merging, I took the latest cloudlib version.

I was not able to update `rsrcd` though. I think at some point `"ticks"` went from an `int` to a `float`, so I just kept my banged up older version.

Said older version has a couple changes because somehow timestamps became (and stayed) negative at one point, causing the loop in the decay function to lag and everything to have issues. I dunno if it fixed anything, but I switched from using `time()` to `status(ST_UPTIME)` and turned the for loop into a `pow` expression, and manually forced all the timestamps to a sane value. It hasn't broken since then, though.

But checking through rsrc, I notice a few resource owners have negative objects, so... hm... though I am doing a few things differently around destructing objects and clones, so I suppose it could be caused by something that I've done.
